testsuite/testsuite-microprofile-health/src/test/java/org/wildfly/swarm/microprofile/health/StartupHealthTests.java

```diff
@@ -0,0 +1,153 @@
+    private Deployer deployer;
+
+    @Before
+    public void before() {
+        client = ClientBuilder.newClient();
+    }
+
+    @After
+    public void after() {
+        if (client != null) {
+            client.close();
+        }
+    }
+
+    @Deployment(name = APP_NAME, managed = false)
+    public static WebArchive deployment() {
+        return ShrinkWrap.create(WebArchive.class)
+            .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml")
+            .addPackage(TestDownReadinessHC.class.getPackage());
+    }
+
+    @Test
+    @InSequence(1)
+    public void testLivenessProbeDuringStartup() throws ExecutionException, InterruptedException {
+        Assert.assertNull(getLiveness());
+
+        Future<Response> checkTask = executor.submit(waitForNonNull(this::getLiveness));
+
+        deployer.deploy(APP_NAME);
+
+        // wait for the first response from the liveness probe
+        Response response = checkTask.get();
+
+        Assert.assertEquals(200, response.getStatus());
+        String entity = response.readEntity(String.class);
+        Assert.assertTrue(entity.contains("UP"));
+        Assert.assertTrue(entity.contains("[]"));
+
+        // wait for the first non-empty response
+        String result = executor.submit(waitForNonNull(() -> {
+            String liveness = getLiveness().readEntity(String.class);
+            return liveness.contains("[]") ? null : liveness;
+        })).get();
+
+        Assert.assertTrue(result.contains("UP"));
+        Assert.assertTrue(result.contains(TestUpLivenessHC.class.getSimpleName()));
+
+        deployer.undeploy(APP_NAME);
+    }
+
+    @Test
+    @InSequence(2)
+    public void testReadinessProbeDuringStartup() throws ExecutionException, InterruptedException {
+        Assert.assertNull(getReadiness());
+
+        Future<Response> checkTask = executor.submit(waitForNonNull(this::getReadiness));
+
+        deployer.deploy(APP_NAME);
+
+        // wait for the first response from the readiness probe
+        Response response = checkTask.get();
+
+        Assert.assertEquals(503, response.getStatus());
+        String entity = response.readEntity(String.class);
+        Assert.assertTrue(entity.contains("DOWN"));
+        Assert.assertTrue(entity.contains("[]"));
+
+        // wait for the first non-empty response
+        String result = executor.submit(waitForNonNull(() -> {
+            String readiness = getReadiness().readEntity(String.class);
+            return readiness.contains("[]") ? null : readiness;
+        })).get();
+
+        Assert.assertTrue(result.contains("DOWN"));
+        Assert.assertTrue(result.contains(TestDownReadinessHC.class.getSimpleName()));
+
+        deployer.undeploy(APP_NAME);
+    }
+
+    private <T> Callable<T> waitForNonNull(Callable<T> waitFunction) {
+        return () -> {
+            T response = null;
+
+            while ((response = waitFunction.call()) == null) {
+                try {
+                    Thread.sleep(1000);
```

Comment:
1 second is a bit too much. It's unlikely, but it can happen that the time between HTTP port becomes available and application deployment finishes is less than 1 second. I'd use 100 millis here.

(Usually, I'd suggest using Awaitility, but for some reason, we don't have that in Thorntail, and it probably doesn't make sense to add it just for this purpose.)